Add `sig_to_mrd` Function for Exporting MRZero Signals to ISMRMRD

felixlandmeyer opened the pull request:
I have written a small function called `sig_to_mrd` that exports simulated MR signals from MRZero to the ISMRMRD file format. This allows me to use the same reconstruction workflows as I do with the data measured on the scanner.

I'm not sure if this is useful for the MRZero platform. If there is interest, I am happy to make further changes.
Feedback is very welcome!

### Summary
- Takes MRZero-simulated signal and corresponding Pulseq sequence.
- Splits the signal into ISMRMRD acquisitions based on ADC IDs and sequence labels.
- Generates a minimal ISMRMRD header using metadata extracted from the `.seq` file.
- Useful for testing reconstructions in a consistent way with real scanner data.

### Tests
- Tested with my custom sequences and the following notebooks:
  - `mr0_diffusion_prep_STEAM_2D_seq.ipynb`
  - `mr0_bSSFP_2D_seq.ipynb`
- May not support all possible sequence configurations.

### Dependencies Added
- [`ismrmrd`](https://github.com/ismrmrd/ismrmrd)
- [`tqdm`](https://github.com/tqdm/tqdm)

mzaiss commented:
Definitly of interest, and on our list since quite some time, so great to have your pull request. (That's a first ISMRM minihub outocme!)

We will review and probably add some discussion here. E.g. Multi coil might be a good addition and is relativeley easy to do.
But might need some coil combination then.

felixlandmeyer commented:
Multichannel should already be possible, I just removed it from the examples in order to not mess up the original notebook too much.

By the way, do you have any predefined coil sensitivity maps for the phantoms which can be automatically loaded together with the phantoms?
At the moment, I am importing coilsens maps from elsewhere.

J-Endres reviewed (commented):
There are lots of formatting changes - while it might be a good idea to apply automatic formatting to this code base in the future, this is not the goal of this pull request and these changes should be removed.

The important changes are in sig_to_mrd.py, correct? Or are there other changes in other files you want to have included, too?

J-Endres commented:
> Multichannel should already be possible, I just removed it from the examples in order to not mess up the original notebook too much.
> 
> By the way, do you have any predefined coil sensitivity maps for the phantoms which can be automatically loaded together with the phantoms? At the moment, I am importing coilsens maps from elsewhere.

There are no multichannel maps included with the provided phantoms.
We might change this in the future for our custom maps but would have to generate those synthetically for BrainWeb - I don't know if that is a good idea and might complicate simple experiments that do not care about this property.

How do you import them, do you have your own simulation data or do you combine it with one of the provided phantoms?

felixlandmeyer commented:
> There are lots of formatting changes - while it might be a good idea to apply automatic formatting to this code base in the future, this is not the goal of this pull request and these changes should be removed.
> 
> The important changes are in sig_to_mrd.py, correct? Or are there other changes in other files you want to have included, too?

Reverted formatting in notebooks and __init__.py.

Important changes are in `sig_to_mad` and dependencies as `ismrmrd` and `tqdm` are added.
Both packages have very permissive licenses.

I have changed the notebooks so that the ismrmrd example is at the end. It is intended more as a ‘Hey, you can do that too if you want’ kind of thing.

felixlandmeyer commented:
> > Multichannel should already be possible, I just removed it from the examples in order to not mess up the original notebook too much.
> > By the way, do you have any predefined coil sensitivity maps for the phantoms which can be automatically loaded together with the phantoms? At the moment, I am importing coilsens maps from elsewhere.
> 
> There are no multichannel maps included with the provided phantoms. We might change this in the future for our custom maps but would have to generate those synthetically for BrainWeb - I don't know if that is a good idea and might complicate simple experiments that do not care about this property.
> 
> How do you import them, do you have your own simulation data or do you combine it with one of the provided phantoms?

I normally use a very simple sensitivity map simulation based on Biot-Savart law, which takes into account different coil channel positions, radii, etc.
While this does not necessarily add any particular value to the simulation, it does enable me to test my reconstruction processes more realistically.

mzaiss commented:
To add ISMRMRD format means to open up to all other recon projects, which means we should definitly add parallel imaging support as everybody does GRAPPA and CS etc.

Thus, we should add some coil sensitivities, best practice similar to what can be measured. So people can simple change a simulated setup and a scanned setup.

Then we should als have an example with multicoil, coil combination and PI recon. But maybe using mri-nufft or mrpro in torch is then possible in the colab.

The simple single coil colab is still nice and should also be kept.

And multi-coil with generated sensitivities is totally fine i think.
